src/utils/upload.js

```diff
@@ -1,88 +1,118 @@
 /**
- * @param {Object} param0
- * @param {Object} param1
- * @returns {Object} file uploader
+ * @param {Object} params
+ * @param {Object} handlers
+ * @param {Object} options
+ * @returns {{
+ *  upload: FormData => {
+ *    id: number,
+ *    request: XMLHttpRequest,
+ *    promise: Promise,
+ *    cancel: () => void
+ *  }
+ * }} file uploader
  */
 function createFileUploader(
   { method, url, accessToken },
   {
     onUploadStart = () => {},
     onUploadProgress = () => {},
     onUploadComplete = () => {},
+    onUploadRetry = () => {},
     onUploadError = () => {},
     onUploadCancel = () => {}
-  }
+  },
+  { retryCount = 0, retryInterval = 0 } = {}
 ) {
-  let lastUploadId = 0;
+  let nextUploadId = 0;
 
   const fileUploader = {
-    request: null,
     upload(data) {
       if (!(data instanceof FormData)) {
         console.error(
-          "[File Uploader] Error: 'data' argument is not an instance of FormData."
+          "[File Uploader] Error: 'data' argument must be an instance of FormData."
         );
         return;
       }
 
-      const uploadId = lastUploadId++;
-      const request = new XMLHttpRequest();
+      let request, resolve, reject;
+      let count = 0;
 
-      request.upload.addEventListener("loadstart", e => {
-        onUploadStart({
-          id: uploadId,
-          bytesUploaded: e.loaded,
-          bytesTotal: e.total,
-          percentage: (e.loaded / e.total) * 100
-        });
-      });
-      request.upload.addEventListener("progress", e => {
-        onUploadProgress({
-          id: uploadId,
-          bytesUploaded: e.loaded,
-          bytesTotal: e.total,
-          percentage: (e.loaded / e.total) * 100
+      const uploadId = nextUploadId++;
+      const promise = new Promise(
+        (res, rej) => ((resolve = res), (reject = rej))
+      );
+
+      const handleError = e => {
+        if (count < retryCount) {
+          count++;
+          onUploadRetry({ id: uploadId, error: e, count });
+          setTimeout(sendRequest, retryInterval);
+        } else {
+          onUploadError({ id: uploadId, error: e });
+          reject({ id: uploadId, error: e });
+        }
+      };
+
+      const sendRequest = () => {
+        // Abort any previous request that is not terminated
+        if (request?.readyState < XMLHttpRequest.DONE) request?.abort();
+
+        request = new XMLHttpRequest();
+
+        request.upload.addEventListener("loadstart", e => {
```

Comment:
just a proposition : 

```js
request.upload.addEventListener("loadstart", ({ loaded: bytesUploaded, total: bytesTotal }) => {
  onUploadStart({
    id: uploadId,
    bytesUploaded,
    bytesTotal,
    percentage: (bytesUploaded / bytesTotal) * 100
  });
});
```

❓

Reply:
Let me come with another proposition (from Hugo):
```js
request.upload.addEventListener("loadstart", ({ loaded, total }) => {
  onUploadStart({
    id: uploadId,
    bytesUploaded: loaded,
    bytesTotal: loaded,
    percentage: (loaded / loaded) * 100
  });
});
```

Reply:
I hope you do not write this in the code ^^ (too much loaded, to total :P )